### ISSUE TITLE
model merge and unload feature for AdaLora

### ISSUE THREAD

**Issue by DaehanKim** [huggingface/peft#2322]
### Feature request

unlike Lora or IA3 adapter type, AdaLora does not provide a method to merge lora adapter weights into original weights so that it can be used as a standalone model. I made that feature for a personal usecase and want to make a PR to make this feature accessible to everyone. 

### Motivation

This feature makes people easily merge AdaLora adapter weights into original weights, which makes further finetuning on it possible (i.e. when one wants to resume adalora training for checkpoints that was already trained with adalora, resuming training is not possible with unmerged weights. )

### Your contribution

I'll submit a PR. I followed the example of IA3 `merge_and_unload`

Following is the overview of change : 

```
    def _unload_and_optionally_merge(
        self,
        merge: bool = True,
        safe_merge: bool = False,
        adapter_names: Optional[list[str]] = None,
        eps: float = 1e-5
    ) -> torch.nn.Module:
        """
        This method unloads the AdaLoRA adapter modules and optionally merges them into the base model weights.
        
        Args:
            merge (`bool`, defaults to `True`):
                If True, merges the adapter weights into base model weights. 
                If False, it will only unload the adapters without merging.
            safe_merge (`bool`, defaults to `False`):
                If True, performs the merge operation with extra safety checks.
            adapter_names (`List[str]`, *optional*):
                The list of adapter names to merge. If None, all active adapters will be merged.
            eps (`float`, defaults to 1e-5):
                Small constant for numerical stability when dividing by ranknum.
                
        Returns:
            model (`torch.nn.Module`):
                The resulting PyTorch model.
        """
        if getattr(self.model, "is_loaded_in_8bit", False):
            raise ValueError("Cannot merge adalora layers when the model is loaded in 8-bit mode")

        if getattr(self.model, "is_loaded_in_4bit", False):
            raise ValueError("Cannot merge adalora layers when the model is loaded in 4-bit mode")
            
        if adapter_names is not None:
            raise ValueError("AdaLoRA does not support merging specific adapters. Got adapter_names={adapter_names}")

        # Create a copy of the base model state dict to modify
        original_state_dict = self.model.state_dict()

        if merge:
            for name, module in self.model.named_modules():
                if hasattr(module, "base_layer") and hasattr(module, "lora_A"):
                    # Extract base layer weight name
                    layer_name = name.replace(".lora_A", "")
                    layer_name = layer_name.replace("base_model.model.", "")
                    base_weight_name = f"{layer_name}.weight"

                    # Get SVD parameters
                    lora_A = module.lora_A["default"]  # [r x d_in]
                    lora_B = module.lora_B["default"]  # [d_out x r]
                    lora_E = module.lora_E["default"]  # [r x 1]
                    
                    # Calculate active ranks
                    ranknum = (lora_E != 0).sum()
                    scaling = module.scaling["default"] if hasattr(module, "scaling") else 16

                    # Safety check if requested
                    if safe_merge and (torch.isnan(lora_A).any() or torch.isnan(lora_B).any() or torch.isnan(lora_E).any()):
                        raise ValueError(f"NaN detected in adapter weights for layer {name}")

                    # Scale A with E: A' = AE
                    scaled_A = lora_A * lora_E  # [r x d_in]

                    # Compute update: ΔW = BA'
                    if ranknum > 0:
                        update = (lora_B @ scaled_A) * scaling / (ranknum + eps)
                    else:
                        update = torch.zeros_like(original_state_dict[base_weight_name])

                    # Update base weights
                    if base_weight_name in original_state_dict:
                        original_state_dict[base_weight_name] += update

        # Load the merged state dict back into a clean version of the model
        self.model.load_state_dict(original_state_dict)

        return self.model

    def merge_and_unload(
        self, 
        safe_merge: bool = False, 
        adapter_names: Optional[list[str]] = None,
        eps: float = 1e-5
    ) -> torch.nn.Module:
        """
        Merge the active adapters into the base model and unload the adapters.
        
        Args:
            safe_merge (`bool`, defaults to `False`):
                If True, performs the merge operation with extra safety checks.
            adapter_names (`List[str]`, *optional*):
                List of adapter names to merge. If None, merges all active adapters.
            eps (`float`, defaults to 1e-5):
                Small constant for numerical stability when dividing by ranknum.
        
        Returns:
            `torch.nn.Module`: The merged model.
        """
        return self._unload_and_optionally_merge(
            safe_merge=safe_merge, 
            adapter_names=adapter_names,
            eps=eps
        )

    def unload(self) -> torch.nn.Module:
        """
        Unload the adapters without merging them into the base model.
        
        Returns:
            `torch.nn.Module`: The unloaded model.
        """
        return self._unload_and_optionally_merge(merge=False)
```

**Comment by BenjaminBossan** [comment]
Thanks for the suggestion. How did you try to merge the AdaLoRA model, as it should already work?

```python
from transformers import AutoModelForCausalLM
from peft import get_peft_model, AdaLoraConfig

model_id = "facebook/opt-125m"
model = AutoModelForCausalLM.from_pretrained(model_id)
model = get_peft_model(model, AdaLoraConfig())
merged = model.merge_and_unload()
```

**Comment by DaehanKim** [comment]
Hi @BenjaminBossan,
I already tried that approach on the firsthand.
But it did not work for Adalora.
Did it work for you?

**Comment by DaehanKim** [comment]
Yes!
As I know, AdaLora weights initializes to contribute nothing to the output logits.  
So your approach would yield same output across untrained base model and untrained merged model.

When trained (i.e. expanded low rank matrix is not zero matrix), Applying `merge_and_unload` only returns base layer weights. Also when I inspect the source of AdaLora, there's no method names with `merge_and_unload`. So my theory is that nothing happens for peft classes that don't have `merge_and_unload` method. Thank you for your attention!
Let me know if I got wrong somewhere.

**Comment by BenjaminBossan** [comment]
We can show that merging AdaLoRA works by extending my snippet from above:

```python
import torch
from transformers import AutoModelForCausalLM
from peft import get_peft_model, AdaLoraConfig

model_id = "facebook/opt-125m"
model = AutoModelForCausalLM.from_pretrained(model_id)
inputs = torch.arange(10).view(-1, 1)

print("logits from base model")
torch.manual_seed(0)
with torch.inference_mode():
    # tensor([-3.9534, -3.9515,  3.2368, -3.9592,  5.4998])
    print(model(inputs).logits[0, 0, :5])

model = get_peft_model(model, AdaLoraConfig(total_step=5))

print("logits from AdaLoRA model before training (should be same as base model)")
torch.manual_seed(0)
with torch.inference_mode():
    # tensor([-7.5773, -7.5917, -1.7851, -7.5924,  3.5000])
    print(model(inputs).logits[0, 0, :5])

# mock training
optimizer = torch.optim.SGD(model.parameters(), lr=0.001)
for epoch in range(5):
    optimizer.zero_grad()
    loss = model(inputs).logits.sum()
    loss.backward()
    optimizer.step()
    model.update_and_allocate(epoch)

print("logits from AdaLoRA model after training (should be different from base model)")
torch.manual_seed(0)
with torch.inference_mode():
    # tensor([-23.7171, -23.7066,   0.9904, -23.7145,  11.1244])
    print(model(inputs).logits[0, 0, :5])

merged = model.merge_and_unload()

print("logits from merged model (should be same as after training)")
torch.manual_seed(0)
with torch.inference_mode():
    # tensor([-23.7171, -23.7066,   0.9904, -23.7145,  11.1244])
    print(merged(inputs).logits[0, 0, :5])
```

Notice that the logits changed after training and that after merging, the logits have the expected values.

> when I inspect the source of AdaLora, there's no method named `merge_and_unload`

`AdaLoraModel` inherits from `LoraModel`, which is where it gets the `merge_and_unload` method from.

**Comment by DaehanKim** [comment]
Thank you for clarifying! I guess I made some mistake when doing merging.
